prov/rxm/src/rxm_cq.c

```diff
@@ -95,30 +95,63 @@ static void rxm_cq_log_comp(uint64_t flags)
 }
 #endif
 
-int rxm_finish_recv(struct rxm_rx_buf *rx_buf)
+static inline int rxm_write_recv_cq(struct rxm_rx_buf *rx_buf)
 {
-	int ret;
-
 	if (rx_buf->recv_entry->flags & FI_COMPLETION) {
 		FI_DBG(&rxm_prov, FI_LOG_CQ, "writing recv completion\n");
-		ret = ofi_cq_write(rx_buf->ep->util_ep.rx_cq,
+		return ofi_cq_write(rx_buf->ep->util_ep.rx_cq,
 				   rx_buf->recv_entry->context,
 				   rx_buf->recv_entry->comp_flags |
 				   (rx_buf->pkt.hdr.flags & OFI_REMOTE_CQ_DATA) ?
 				   FI_REMOTE_CQ_DATA : 0,
 				   rx_buf->pkt.hdr.size, NULL,
 				   rx_buf->pkt.hdr.data, rx_buf->pkt.hdr.tag);
-		if (ret) {
-			FI_WARN(&rxm_prov, FI_LOG_CQ,
-					"Unable to write recv completion\n");
-			return ret;
-		}
+	} else {
+		return FI_SUCCESS;
+	}
+}
+
+static inline int rxm_finish_recv(struct rxm_rx_buf *rx_buf)
+{
+	int ret = rxm_write_recv_cq(rx_buf);
+	if (OFI_UNLIKELY(ret)) {
+		FI_WARN(&rxm_prov, FI_LOG_CQ,
+			"Unable to write recv completion\n");
+		return ret;
 	}
 
 	rxm_recv_entry_release(rx_buf->recv_queue, rx_buf->recv_entry);
+
 	return rxm_ep_repost_buf(rx_buf);
 }
 
+static inline int rxm_finish_recv_ex(struct rxm_rx_buf *rx_buf)
+{
+	int ret = rxm_write_recv_cq(rx_buf);
+	if (OFI_UNLIKELY(ret)) {
+		FI_WARN(&rxm_prov, FI_LOG_CQ,
+			"Unable to write recv completion\n");
+		return ret;
+	}
+
+	if (++rx_buf->ep->rx_buf_cnt == rx_buf->ep->max_rx_buf_cnt) {
+		struct rxm_rx_buf *repost_rx_buf;
+		while (!dlist_empty(&rx_buf->ep->rx_buf_list)) {
+			dlist_pop_front(&rx_buf->ep->rx_buf_list, struct rxm_rx_buf,
+					repost_rx_buf, entry);
+			rxm_recv_entry_release(repost_rx_buf->recv_queue,
+					       repost_rx_buf->recv_entry);
+			(void)rxm_ep_repost_buf(repost_rx_buf);
+		}
+		rx_buf->ep->rx_buf_cnt = 0;
+		rxm_recv_entry_release(rx_buf->recv_queue, rx_buf->recv_entry);
+		return rxm_ep_repost_buf(rx_buf);
+	} else {
+		dlist_insert_tail(&rx_buf->entry, &rx_buf->ep->rx_buf_list);
+		return FI_SUCCESS;
+	}
```

Comment:
If I understand correctly, rather than reposting receive buffer immediately, we instead put it on a deferred post queue.  Once the deferred queue reaches a certain threshold, we then post them.  I don't see how this helps overall.  Obviously inserting on a deferred queue is faster than reposting it.  But we'll eventually need to repost the buffer, in which case, we've ended up doing more work -- adding queue plus dequeue.

I understand why verbs might do this to save PCI transactions.  But without RxM going directly to verbs here (which I think is a reasonable option), I don't think we'll see any benefit.

Reply:
Yes. you're right.
I've dropped the changes related to re-posting buffers for now.